docs/2019-09-25-Applying-Axioms-By-Matching.md

```diff
@@ -0,0 +1,165 @@
+Algorithm for applying one rule to a pattern by matching
+========================================================
+
+This document presents an alternative way of applying axioms,
+both rewrite- and equality- ones, which can be used when
+the rule fully matches the configuration, thus having the advantage
+that it does not produce remainders.
+
+Since it makes no assumptions on the nature of the patterns involved,
+it can safely be used when the configuration or the matching pattern
+are __not__ function-like.
+
+Rewrite rules
+-------------
+
+Given a rewrite axiom of the form 
+(0) `∀ Y .α(Y) → •β(Y)`
+and a goal `φ → •ψ` such that 
+(1) `φ` can be decomposed as `φ = φₜ ∧ φₚ`, (1') `φₚ` predicate.
+(2) `α(Y)` can be decomposed as `α(Y) = αₜ(Y) ∧ αₚ(Y)`
+and there exists `t` such that
+(3) `φₚ → (φₜ = αₜ(t))` and
+(4) `φₚ → αₚ(t)`
+
+Then, if we can show that
+(5) `β(t) ∧ φₚ → ψ`
+
+Then it is true that `φ → •ψ`.
+
+### Proof
+
+
+Premise (0) says
+`∀ Y .α(Y) → •β(Y)`; instantiating `Y` with `t` we get
+`α(t) → •β(t)` which is equivalent by (2) with
+(6) `αₜ(t) ∧ αₚ(t) → •β(t)`.
+
+From (3) and (4), by FOL reasoning (conclusions with same premise), we get
+`φₚ → (φₜ = αₜ(t)) ∧ αₚ(t)`; by FOL reasoning (conjunction to left), we get
+`φₜ ∧ φₚ → φₜ ∧ (φₜ = αₜ(t)) ∧ αₚ(t)`; by equality elimination we get
+`φₜ ∧ φₚ → αₜ(t) ∧ αₚ(t)`; by (6) and transitivity of `→`, we get
+`φₜ ∧ φₚ → •β(t)`; by FOL reasoning we get
+`φₜ ∧ φₚ → •β(t) ∧ φₚ`; due to (1') and properties of conjunction we get
+`φₜ ∧ φₚ → •(β(t) ∧ φₚ)`, which is equivalent by (1) with
+(7) `φ → •(β(t) ∧ φₚ)`.
+
+Premise (5) says
+`β(t) ∧ φₚ → ψ`; by the Framing rule of ML with `•`, we get
+`•(β(t) ∧ φₚ) → •ψ`; by (7) and transitivity of `→`, we get
+`φ → •ψ`.
+
+QED
+
+### Coverage concerns (All-path-reachability)
+
+Note that if `φ` and `αₜ(Y)` are not constructor-based
+(in which case the unifier `Y = t` is an MGU) then 
```

Comment:
As discussed:
* I think that it's hard to parse "in which case" as referring to "is constructor-based".
* I think that the constructors must be functional.
* Any injective symbol may work, not only constructors.